plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm

```diff
@@ -0,0 +1,201 @@
+{
+	blog(LOG_DEBUG, "output_start");
+
+	[sMachServer run];
+
+	obs_get_video_info(&videoInfo);
+
+	struct video_scale_info conversion = {};
+	conversion.format = VIDEO_FORMAT_UYVY;
+	conversion.width = videoInfo.output_width;
+	conversion.height = videoInfo.output_height;
+	obs_output_set_video_conversion(output, &conversion);
+	if (!obs_output_begin_data_capture(output, 0)) {
+		return false;
+	}
+
+	return true;
+}
+
+static void virtualcam_output_stop(void *data, uint64_t ts)
+{
+	blog(LOG_DEBUG, "output_stop");
+	obs_output_end_data_capture(output);
+	[sMachServer stop];
+}
+
+static void virtualcam_output_raw_video(void *data, struct video_data *frame)
+{
+	uint8_t *outData = frame->data[0];
+	if (frame->linesize[0] != (videoInfo.output_width * 2)) {
+		blog(LOG_ERROR,
+		     "unexpected frame->linesize (expected:%d actual:%d)",
+		     (videoInfo.output_width * 2), frame->linesize[0]);
+	}
+
+	CGFloat width = videoInfo.output_width;
+	CGFloat height = videoInfo.output_height;
+
+	[sMachServer sendFrameWithSize:NSMakeSize(width, height)
+			     timestamp:frame->timestamp
+			  fpsNumerator:videoInfo.fps_num
+			fpsDenominator:videoInfo.fps_den
+			    frameBytes:outData];
+}
+
+static bool GetDALPluginStatus()
+{
+	NSFileManager *fileManager = [NSFileManager defaultManager];
+
+	NSString *pathForFile =
+		@"/Library/CoreMediaIO/Plug-Ins/DAL/obs-mac-virtualcam.plugin";
+
+	return [fileManager fileExistsAtPath:pathForFile];
+}
+
+static void CopyDALPlugin()
+{
+	NSString *sourcePath;
+	NSRunningApplication *app = [NSRunningApplication currentApplication];
+
+	if ([app bundleIdentifier] != nil) {
+		NSURL *bundleURL = [app bundleURL];
+		NSString *pluginPath =
+			@"Contents/Resources/data/obs-mac-virtualcam.plugin";
+
+		NSURL *pluginUrl =
+			[bundleURL URLByAppendingPathComponent:pluginPath];
+		sourcePath = [pluginUrl path];
+	} else {
+		sourcePath = [[[[app executableURL]
+			URLByAppendingPathComponent:
+				@"../data/obs-mac-virtualcam.plugin"] path]
+			stringByReplacingOccurrencesOfString:@"obs/"
+						  withString:@""];
+	}
+
+	NSFileManager *fileManager = [NSFileManager defaultManager];
+
+	if ([fileManager fileExistsAtPath:sourcePath]) {
+		NSString *destinationPath =
+			@"/Library/CoreMediaIO/Plug-Ins/DAL/obs-mac-virtualcam.plugin";
+		NSString *copyCmd = [NSString
+			stringWithFormat:
+				@"do shell script \"cp -R '%@' '%@'\" with administrator privileges",
+				sourcePath, destinationPath];
+
+		NSDictionary *errorDict;
+		NSAppleEventDescriptor *returnDescriptor = NULL;
+		NSAppleScript *scriptObject =
+			[[NSAppleScript alloc] initWithSource:copyCmd];
+		returnDescriptor =
+			[scriptObject executeAndReturnError:&errorDict];
+		if (errorDict != nil) {
+			const char *errorMessage = [[errorDict
+				objectForKey:@"NSAppleScriptErrorMessage"]
+				UTF8String];
+			blog(LOG_INFO,
+			     "[macOS] VirtualCam DAL Plugin Installation status: %s",
+			     errorMessage);
+		}
+	} else {
+		blog(LOG_INFO,
```

Comment:
Idea: I wonder if we should detect this up front and not show the menu item at all. Though if we suspect this will normally work, it's probably not worth worrying about.

Reply:
Yeah there might be more elegant variants for this, but my train of thought was this:

* How to get this into the codebase quickest to minimise the gap of time between shipping the feature on Windows and macOS?
* How long until we get the permission onboarding flow (which would include the button to trigger this installation)?
* How to re-enable this functionality if users dismiss the onboarding screen/forget to install the DAL plugin?
* How not to contribute to the amount of prompts happening at first OBS launch already?

I've found that installing it on demand for now using the menu item was quickest and easiest. If we hide the menu item otherwise, we'd need to find a home for the installation button, which would also introduce a dependency to the main UI codebase.

Hence my idea to solve it like this for now and handle it more gracefully as part of the global macOS permission solution later.

Reply:
Yeah thank you for just knocking it out! Reflecting on it a bit, I think I had been procrastinating because I wasn't sure how to build the permission onboarding flow. I think your approach to get something we can ship and then improve it in the future is right.

This solution is fine by me. I'll leave it to @jp9000 and @dodgepong to determine if it's shippable.